docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
+Enabling PSR-4 auto-loading is optional. It will provide better auto-loading of your classes in your development environment and will future proof your code for SilverStripe 5.
 
-* PHP 5.6 is now the minimum required version (and up to PHP 7.1.x is supported!).
-* CMS CSS has been re-developed using Bootstrap 4 as a base.
-* SilverStripe code _should_ now be [PSR-2 compliant](http://www.php-fig.org/psr/psr-2/). While this is not a requirement, we strongly suggest you switch over now. You can use tools such as [`phpcbf`](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Fixing-Errors-Automatically) to do most of it automatically.
-* We've also introduced some best practices for module development. [See the Modules article](/developer_guides/extending/modules) for more information.
+For example, let's say you have defined the following namespaces for the following folders:
+* `Acme\Web` for your main application logic contained in `mysite/code`
+* `Acme\SubModule` for a secondary module contained in `sub-module/code`
+* `Acme\Web\Tests` for your application test suite contained in `mysite/tests`.
+
+Your `autoload` section in your `composer.json` file should look like this:
+```json
+{
+    // ...
+    "autoload": {
+        "psr-4": {
+            "Acme\\Web\\": "mysite/code",
+            "Acme\\SubModule\\": "sub-module/code"
+        }
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "Acme\\Web\\Tests\\": "mysite/tests"
+        }
+    },
+    // ...
+}
+```
+
+Read the [Composer schema autoload documentation](https://getcomposer.org/doc/04-schema.md#autoload) for more information about configuring auto-loading in your project.
+
+### Finalise your project namespacing
+All your classes should now be fully namespaced.
+
+Note that applying a namespace to your project will also affect which template file SilverStripe tries to load when rendering certain objects.
+
+For example, pretend you have a `RoadRunnerPage` class that extends `Page`. In SilverStripe 3, you would define a template for this page in `themes/acme/templates/Layout/RoadRunnerPage.ss`. If you decide to move `RoadRunnerPage` to `Acme\Web\RoadRunnerPage`, you'll need to move the template to `themes/acme/templates/Acme/Web/Layout/RoadRunnerPage.ss`.
+
+This is a good point to commit your changes to your source control system before moving on to the next step.
+
+## Step 4 - Update codebase with references to newly namespaced classes {#step4}
+
+Nearly all core PHP classes in SilverStripe 4 have been namespaced. For example, `DataObject` is now called `SilverStripe\ORM\DataObject`. Your project codebase, config files and language files need be updated to reference those newly namespaced classes. This will include explicit references in your PHP code, but also string that contain the name of a class.
```

Comment:
That bit got copied over, so not quite sure what nearly refers to in this context. 

There's some weird things under `thirdparty` that aren't namespaced, but I wouldn't call those core classes.

All drop the nearly to avoid confusing people.